### PR TITLE
Use `DeclRef` for struct and enum TypeInfo

### PR DESCRIPTION
**Pull request by IGI-111**
## Description

Use references and the decl engine to resolve information about structs and enums instead of copying around declaration data.

This should facilitate #3744

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Comment by emilyaherbert** [comment]
Thanks @IGI-111! Awesome work on this.

**Comment by emilyaherbert** [comment]
CC @eureka-cpu, this PR should unblock you.

File: sway-core/src/type_system/info.rs [hunk 1]
```diff
@@ -92,16 +93,8 @@ pub enum TypeInfo {
...
-        call_path: CallPath,
-        type_parameters: Vec<TypeParameter>,
-        variant_types: Vec<ty::TyEnumVariant>,
-    },
-    Struct {
-        call_path: CallPath,
-        type_parameters: Vec<TypeParameter>,
-        fields: Vec<ty::TyStructField>,
-    },
+    Enum(DeclRefEnum),
```
Comment (eureka-cpu) [inline sway-core/src/type_system/info.rs:96]: Nice, excited for this to go through! 🚀

File: test/src/e2e_vm_tests/harness.rs [hunk 2]
```diff
@@ -299,6 +299,7 @@ pub(crate) fn test_json_abi(file_name: &str, built_package: &BuiltPackage) -> Re
         fs::read_to_string(oracle_path).expect("Something went wrong reading the file.");
     let output_contents =
         fs::read_to_string(output_path).expect("Something went wrong reading the file.");
+    dbg!(&oracle_contents, &output_contents);
```
[the remark this answers is not in the document]
Comment (IGI-111) [reply test/src/e2e_vm_tests/harness.rs:302]: Nope :sweat_smile:

**Review by mohammadfawaz — APPROVED** [review]

**Pull request merged**